generate_latest_json.py

```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+"""
+Sopel update-check JSON generation utility
+
+This script creates a `latest.json` file representing the most recent stable and
+unstable `sopel` releases published to PyPI, using PyPI's own JSON API.
+
+Copyright 2024 dgw, technobabbl.es
+Licensed under the Eiffel Forum License 2.
+
+https://sopel.chat
+"""
+import json
+
+from packaging.version import parse as parse_version
+import requests
+
+
+def main(argv=None):
+    data = requests.get('https://pypi.org/pypi/sopel/json').json()
+    releases = sorted([parse_version(v) for v in data['releases'].keys()])
+
+    latest = releases[-1]
+    stable = latest
+    if latest.is_prerelease:
+        for v in reversed(releases[:-1]):
+            if not v.is_prerelease:
+                stable = v
+                break
+        else:
+            print("ERROR: Latest version is a prerelease, but no earlier stable version found!")
+            exit(1)
+
+    with open('latest.json', 'w') as f:
+        f.write("{{\n"  # double braces so Python doesn't get confused
+                "    \"version\": \"{stable}\",\n"
+                "    \"unstable\": \"{latest}\",\n"
+                "    \"release_notes\": \"https://sopel.chat/changelog/{stable}/\",\n"
+                "    \"unstable_notes\": \"https://github.com/sopel-irc/sopel/releases/v{latest}\"\n"
+                "}}".format(latest=latest, stable=stable))
+
+    print("Generated latest.json file pointing to {} (stable) + {} (unstable)."
+          .format(stable, latest))
```

Comment:
I think this message is lying in the case where `latest` _is_ stable? In particular, when `stable == latest`, it will produce a message that claims that the version in question is both stable and unstable. Low risk, but I feel obliged to point it out :stuck_out_tongue: 

As I think about that, I think the new field named `unstable_notes` in the JSON file may have the same issue?

Reply:
Like so many other things in the Sopel world, these field names are an inheritance. 😅 The "unstable" label seems to be more of a "release channel" than a description of the particular release.

If you look at [the upstream `find_updates` logic](https://github.com/sopel-irc/sopel/blob/1237568f05a4155f80e12ff0fd3711e8ed97d976/sopel/builtins/find_updates.py#L93-L107) it's a bit clearer: The `unstable` and `unstable_notes` fields are used if the Sopel version doing the checking is an unstable release. It'll never prompt the admin of a bot running the latest stable version to update to a newer unstable release, but anyone running an unstable version older than the latest stable release will get prompted.

The only way to get on the unstable "channel" right now is to manually install a prerelease, but once someone does that they should be notified when the unstable "channel" is actually ready to be replaced by a stable version.

Does that make sense? I'm rushing a bit to write all this before my evening obligations start 😁

Reply:
Yea, fair enough!